wallet/src/de/schildbach/wallet/service/WalletTransactionMetadataProvider.kt

```diff
@@ -102,20 +116,185 @@ class WalletTransactionMetadataProvider @Inject constructor(
+                                ScriptPattern.extractHashFromP2SH(output.scriptPubKey)
+                            )
+                        else -> null // for now ignore OP_RETURN (DashPay Expense?)
+                    }
+                    if (address != null) {
+                        val addressMetadata =
+                            addressMetadataDao.loadRecipient(address.toBase58())
+                        if (addressMetadata != null) {
+                            taxCategory = addressMetadata.taxCategory
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return taxCategory
     }
 
     override suspend fun getAllTransactionMetadata(): List<TransactionMetadata> {
-        return transactionMetadataDao.load()
+        val metadataList = transactionMetadataDao.load()
+
+        // look up the transaction metadata and merge with address metadata
+        for (metadata in metadataList) {
+            if (metadata.taxCategory == null) {
+                // if there is no user specified tax category, then look at address_metadata
+                val taxCategory = getDefaultTaxCategory(metadata.txid)
+
+                if (taxCategory != null) {
+                    metadata.taxCategory = taxCategory
+                }
+            }
+        }
+        return metadataList
     }
 
+    override suspend fun markAddressWithTaxCategory(
+        address: String,
+        sendTo: Boolean,
+        taxCategory: TaxCategory
+    ) {
+        // check to see if this address has been used before
+        // if it has been used before, that means the same address was used more than once
+        // possibly for two different services or actions.
+        // This may not matter as the default tax category is probably the same for each.
+        if (addressMetadataDao.exists(address, sendTo)) {
+            log.info("address $address/$sendTo was already added")
+        }
+
+        addressMetadataDao.markAddress(address, sendTo, taxCategory)
+    }
 
+    override suspend fun maybeMarkAddressWithTaxCategory(
+        address: String,
+        sendTo: Boolean,
+        taxCategory: TaxCategory
+    ) {
+        if (!addressMetadataDao.exists(address, sendTo)) {
+            addressMetadataDao.markAddress(address, sendTo, taxCategory)
+        }
+    }
+
+
+    @OptIn(ExperimentalCoroutinesApi::class)
     override fun observeTransactionMetadata(txId: Sha256Hash): Flow<TransactionMetadata?> {
-        return transactionMetadataDao.observe(txId)
+        return transactionMetadataDao.observe(txId).map { transactionMetadata ->
+            // if there is no user specified tax category, then look at address_metadata
+            if (transactionMetadata != null && transactionMetadata.taxCategory == null) {
+                transactionMetadata.taxCategory = getDefaultTaxCategory(txId)
+            }
+            transactionMetadata
+        }
```

Comment:
as in the other functions in this class, before returning TransactionMetadata, we need to check the address table if no tax category is specified.

Currently, transactionMetadataDao is only storing user-specified tax categories.  If the user has not chosen a tax category, then a default value is taken from the addressMetadataDao table if the transaction has addresses that line up with that table.

I did wonder if it would be better to save this default in the transaction metadata table.